core/handler/device/store.go

```diff
@@ -23,11 +24,13 @@ type Store interface {
 	DownlinkQueue(appID, devID string) (DownlinkQueue, error)
 	Set(new *Device, properties ...string) (err error)
 	Delete(appID, devID string) error
+	AttrWhitelist() map[string]bool
 }
 
 const defaultRedisPrefix = "handler"
 const redisDevicePrefix = "device"
 const redisDownlinkQueuePrefix = "downlink"
+const redisUpStreamMetadataKeyWhitelist = ":attr-whitelist"
```

Comment:
2. Yes
4. Maybe you can already limit the number of custom attributes when getting them out of the data store? This is probably also the same on repeats. What I'd like to avoid is that if people have more custom attributes than we allow to publish upstream, that on each message the attributes are different (e.g. if we use a `map[string]string` and break the range after n, the result is random)